File: v2/src/main/mongo/registry.ts
import { MongoClient, type Document } from 'mongodb';
import type { ConnectionRecord } from '../../shared/connection.js';
import type { ConnectionState, TestResult } from '../../shared/connectionState.js';
import { openTunnel, rewriteUriToLocal, type ActiveTunnel } from './tunnel.js';

interface ActiveConnection {
  client: MongoClient;
  tunnel?: ActiveTunnel;
}

type StateListener = (id: string, state: ConnectionState) => void;

export class MongoRegistry {
  private readonly clients = new Map<string, ActiveConnection>();
  private readonly states = new Map<string, ConnectionState>();
  private readonly listeners = new Set<StateListener>();

  onStateChange(listener: StateListener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  getState(id: string): ConnectionState {
    return this.states.get(id) ?? { kind: 'disconnected' };
  }

  snapshot(): ReadonlyMap<string, ConnectionState> {
    return new Map(this.states);
  }

  async connect(record: ConnectionRecord): Promise<ConnectionState> {
    if (this.clients.has(record.id)) {
      return this.getState(record.id);
    }
    this.setState(record.id, { kind: 'connecting' });

    let tunnel: ActiveTunnel | undefined;
    let client: MongoClient | undefined;

    try {
      let uri = record.uri;
      if (record.ssh) {
        const target = parseFirstHostPort(uri);
        tunnel = await openTunnel(record.ssh, target.host, target.port);
        uri = rewriteUriToLocal(uri, tunnel.localHost, tunnel.localPort);
      }

      client = new MongoClient(uri, {
        serverSelectionTimeoutMS: 8_000,
        connectTimeoutMS: 8_000,
      });
      await client.connect();

      const info = await probeServer(client);
      this.clients.set(record.id, { client, tunnel });
      const state: ConnectionState = {
        kind: 'connected',
        pingMs: info.pingMs,
        serverVersion: info.version,
        topology: info.topology,
        connectedAt: Date.now(),
      };
      this.setState(record.id, state);
      return state;
    } catch (err) {
      await safeCleanup(client, tunnel);
      const state: ConnectionState = {
        kind: 'error',
        message: err instanceof Error ? err.message : String(err),
      };
      this.setState(record.id, state);
      return state;
    }
  }

  async disconnect(id: string): Promise<void> {
    const active = this.clients.get(id);
    if (!active) return;
    this.clients.delete(id);
    await safeCleanup(active.client, active.tunnel);
    this.setState(id, { kind: 'disconnected' });
  }

  async disconnectAll(): Promise<void> {
    await Promise.allSettled([...this.clients.keys()].map((id) => this.disconnect(id)));
  }

  async test(record: ConnectionRecord): Promise<TestResult> {
    let tunnel: ActiveTunnel | undefined;
    let client: MongoClient | undefined;
    try {
      let uri = record.uri;
      if (record.ssh) {
        const target = parseFirstHostPort(uri);
        tunnel = await openTunnel(record.ssh, target.host, target.port);
        uri = rewriteUriToLocal(uri, tunnel.localHost, tunnel.localPort);
      }
      client = new MongoClient(uri, {
        serverSelectionTimeoutMS: 6_000,
        connectTimeoutMS: 6_000,
      });
      await client.connect();
      const info = await probeServer(client);
      return {
        ok: true,
        pingMs: info.pingMs,
        serverVersion: info.version,
        topology: info.topology,
      };
    } catch (err) {
      return { ok: false, error: err instanceof Error ? err.message : String(err) };
    } finally {
      await safeCleanup(client, tunnel);
    }
  }

  getClient(id: string): MongoClient | undefined {
    return this.clients.get(id)?.client;
  }

  private setState(id: string, state: ConnectionState): void {
    this.states.set(id, state);
    for (const l of this.listeners) l(id, state);
  }
}

interface ServerInfo {
  pingMs: number;
  version: string;
  topology: string;
}

async function probeServer(client: MongoClient): Promise<ServerInfo> {
  const admin = client.db('admin');
  const t0 = performance.now();
  await admin.command({ ping: 1 });
  const pingMs = Math.round(performance.now() - t0);

  const buildInfo = (await admin.command({ buildInfo: 1 })) as Document;
  const isMaster = (await admin.command({ hello: 1 })) as Document;

  const version = String(buildInfo['version'] ?? 'unknown');
  const topology = describeTopology(isMaster);
  return { pingMs, version, topology };
}

function describeTopology(hello: Document): string {
  if (hello['msg'] === 'isdbgrid') return 'sharded';
  if (typeof hello['setName'] === 'string') return `replica-set (${hello['setName']})`;
  if (hello['isWritablePrimary'] === true || hello['ismaster'] === true) return 'standalone';
  return 'unknown';
}

function parseFirstHostPort(uri: string): { host: string; port: number } {
  const parsed = new URL(uri);
  const host = parsed.hostname || '127.0.0.1';
  const port = parsed.port ? Number(parsed.port) : 27017;
  return { host, port };
}

async function safeCleanup(client: MongoClient | undefined, tunnel: ActiveTunnel | undefined): Promise<void> {
  if (client) {
    try {
      await client.close(true);
    } catch {
      /* swallow */
    }
  }
  if (tunnel) {
    try {
      await tunnel.close();
    } catch {
      /* swallow */
    }
  }
}

File: v3/src/main/kotlin/io/mex/mongo/Registry.kt
package io.mex.mongo

import com.mongodb.ConnectionString
import com.mongodb.MongoClientSettings
import com.mongodb.client.MongoClient
import com.mongodb.client.MongoClients
import io.mex.data.ConnectionRecord
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.withContext
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

class MongoRegistry {
    private val clients = ConcurrentHashMap<String, MongoClient>()
    private val _states = MutableStateFlow<Map<String, ConnectionState>>(emptyMap())
    val states: StateFlow<Map<String, ConnectionState>> = _states.asStateFlow()

    fun state(id: String): ConnectionState = _states.value[id] ?: ConnectionState.Disconnected

    fun client(id: String): MongoClient? = clients[id]

    suspend fun connect(record: ConnectionRecord): ConnectionState =
        withContext(Dispatchers.IO) {
            clients[record.id]?.let { return@withContext state(record.id) }
            setState(record.id, ConnectionState.Connecting)
            try {
                val client = buildClient(record.uri)
                val info = probeServer(client)
                clients[record.id] = client
                val state = ConnectionState.Connected(
                    pingMs = info.pingMs,
                    serverVersion = info.version,
                    topology = info.topology,
                    connectedAt = System.currentTimeMillis(),
                )
                setState(record.id, state)
                state
            } catch (e: Exception) {
                ConnectionState.Error(e.message ?: e::class.java.simpleName).also {
                    setState(record.id, it)
                }
            }
        }

    suspend fun disconnect(id: String): Unit = withContext(Dispatchers.IO) {
        clients.remove(id)?.runCatching { close() }
        setState(id, ConnectionState.Disconnected)
    }

    suspend fun disconnectAll(): Unit = withContext(Dispatchers.IO) {
        clients.keys.toList().forEach { disconnect(it) }
    }

    suspend fun test(uri: String): TestResult = withContext(Dispatchers.IO) {
        var client: MongoClient? = null
        try {
            client = buildClient(uri)
            val info = probeServer(client)
            TestResult(
                ok = true,
                pingMs = info.pingMs,
                serverVersion = info.version,
                topology = info.topology,
            )
        } catch (e: Exception) {
            TestResult(ok = false, error = e.message ?: e::class.java.simpleName)
        } finally {
            client?.runCatching { close() }
        }
    }

    private fun buildClient(uri: String): MongoClient {
        val settings = MongoClientSettings.builder()
            .applicationName("mongo-explorer-v3")
            .applyConnectionString(ConnectionString(uri))
            .applyToClusterSettings { it.serverSelectionTimeout(8, TimeUnit.SECONDS) }
            .applyToSocketSettings { it.connectTimeout(8, TimeUnit.SECONDS) }
            .build()
        return MongoClients.create(settings)
    }

    private fun setState(id: String, state: ConnectionState) {
        _states.update { it + (id to state) }
    }
}
